### DISCUSSION TITLE
No option to pass through existing env vars as well as auto-reload on env change

### DISCUSSION THREAD

**Discussion by martenjacobs** [laravel/framework#43489] [Q&A]
https://github.com/laravel/framework/blob/e7ac9b1e8a20ccb55afe05954f13bb48bb3f8dad/src/Illuminate/Foundation/Console/ServeCommand.php#L135-L139

I'm running Laravel sail in a devcontainer. I'd like to be able to pass through most environment variables to multiple containers so that I can set up MySQL with certain credentials, and use the same env vars in Laravel to connect to the database, for instance. However, I found that serving Laravel through `php artisan serve` will block most environment variables from the "outside" environment. This means I'd have to duplicate all my variables to the .env file for my Laravel app, or use "--no-reload", in which case the filter is ignored. Wouldn't it be possible to have both?

**Comment by JonathanGawrych** [comment]
After hours of debugging why serve via sail didn't respect the DB docker-compose.yml environment variables I set, I came across this.
I echo the confusion expressed in https://github.com/laravel/framework/commit/700b277ebfc5862128358bcdb2dc68edcbeb4bea#commitcomment-53272381 (a comment on where the change was introduced).

I fixed it by adding to my bootstrap/app.php:
```php
ServeCommand::$passthroughVariables[] = 'DB_HOST';
ServeCommand::$passthroughVariables[] = 'DB_PORT';
```
And added warning comments in both here and docker-compose.yml to keep the list in sync.

Then I realized that it used loose comparison `in_array`, so we can use a hack to allow anything (well, besides empty string or '0' but you won't have .env variable keys like that):
```php
ServeCommand::$passthroughVariables[] = true;
```

I don't really know the downside of this... The only thing I can think of is potential confusion on a developer's part if the change the `.env` expecting the serve command to reload with a value, but it doesn't change because the system ENV still overrides the `.env`. However, I'd argue that only _some_ system envs be respected _only_ in serve is much more confusing. Especially since tinker will give different values. So I guess I'll add this to all laravel projects from now. If I find another downside I'll come back and amend my comment.

**Comment by jpswade** [comment]
> Any variable in your `.env` file can be overridden by external environment variables such as server-level or system-level environment variables.

Ref: https://github.com/laravel/docs/blob/83b872eb2e81df9bead2eef7155bbc42b78c55b8/configuration.md?plain=1#L55

Sadly, this isn't true if you're using the serve command, so you need to allow the passthroughVariables as per above:

```
ServeCommand::$passthroughVariables[] = true;
```

This allows DB_HOST, etc to be passed through from the environment to PHP and therefore Laravel.

Perhaps the docs should be updated to reflect this.
